### ISSUE TITLE
Improve authorizer performance

### ISSUE THREAD

**Issue by sbernauer** [stackabletech/hdfs-utils#48]
In [Discord](https://discord.com/channels/796665978481803304/1298956868672163870/1298956877970935819) it was reported

> We are running  hdfs inside Stackable with Kerberos+OPA. The CPU usage looks strange to me and the performance, even for for simple file operations (find, ls, du) is awful. While running spark jobs with data access we see consistent. high CPU load on the Namenode and OPA: 
>  (197.0 % opa, 130.1% Namenode) in top on the host machine. 
> Even setting "default allow true" in OPA and removing everything else does nothing. (basically disabling OPA) 

> I can trigger this with a simplehdfs dfs -du hdfs://our-staackable-hdfs/data/year=2024/month=10
> This takes 50 seconds (while idle) for this to run. (9200 files in 2300 dirs.) 
> And every time ~50 seconds, which leads me to belive that there is zero caching anywhere...
> For comparison: Our seven year old server in the old cluster does this query in 2 seconds.

We believe there are multiple ways this can be improved and we implemented those in these PRs:

- [x] https://github.com/stackabletech/hdfs-utils/pull/50
- [x] https://github.com/stackabletech/hdfs-utils/pull/49

## Make requests to OPA smaller

One reason for the slowness can be because the JSON requests send between HDFS and OPA are gigantic. We need to somehow reduce this size.

Without having thought too long about this, one solution could be to not send the `inodeAttrs`, `inodes`, `pathByNameArr` and `callerContext` fields by default, while still having a opt-in config for users, so that users that care about this fields can still get them.

## Misc improvements

The other prbolem we found (well @iAlex97 did, thank you 📯 ) was that the Enforce would be created over and over again, creating thread  pools and unnecessary Jackson Mappers etc.
In other words: There are some easy wins to be made by improving the Java code.

**Comment by jradmacher** [comment]
I just checked a json-dump of an OPA-request I had lying around: 4348 lines (pretty-json) and 150kb. 99% is `inodeAttrs`, `inodes` and `pathByNameArr` . Your are right, this is huge.

**Comment by iAlex97** [comment]
Jumping in here, I had a similar experience when starting some real workloads which interact with HDFS and need authz. Apart from stripping the JSON which is sent to OPA, I think there is an easy improvement to be made.

In my tests, by turning on logging in the constructor of `StackableAccessControlEnforcer.java` we can see it is instantiated multiple times, hence the Jackson `ObjectMapper` from [here](https://github.com/stackabletech/hdfs-utils/blob/55598e44cd6cf46a3b44c6cb04bab63ede1b6a65/src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java#L66) is recommended to be made static. I cannot find the original post which I stumbled upon when researching this, but [this](https://stackoverflow.com/a/57671444/2512380) answer from SO is pretty similar.

**Comment by iAlex97** [comment]
Another wired thing which we experienced when enabling OPA and NNs were under load was them crashing out once every 15-20 minutes due to "too many files open". Connecting VisualVM to the NN, saw an ever increasing number of threads.

![Image](https://github.com/user-attachments/assets/a44109cd-db40-4170-8357-4d599c1a50ec)
![Image](https://github.com/user-attachments/assets/6a28e1b7-8050-4bda-940f-e2f62b9fdf74)

Those threads seem to be created from the HttpClient under `StackableAccessControlEnforcer.java` and while they should normally be released after the HttpClient is no longer used, or at least they were getting created much faster than reclaimed.

AFAIK HttpClient should be thread-safe and would be quite ok to reuse across different instances of `StackableAccessControlEnforcer`. While it's not specified anywhere in the docs, this [introduction](https://openjdk.org/groups/net/httpclient/intro.html) of the HttpClient implies multi-threading:
> Once built, an HttpClient can be used to send multiple requests.

The only caveat would be that, while HttpClient is thread-safe, apparently SSLContext isn't as mentioned in this [post](https://stackoverflow.com/a/53767728/2512380). While it isn't a problem for now, it's worth keeping in mind if you were to use https instead of plain http.

Therefore, by creating a static HttpClient along with using an executor with fixed thread pool or cached thread pool, I was able keep the thread number pretty stable:
![Image](https://github.com/user-attachments/assets/e35a75a2-2891-431e-93cf-0431ab765e32)

The change is pretty minor:
```java
    private final static HttpClient httpClient = HttpClient.newBuilder()
            .executor(Executors.newCachedThreadPool())
            .build();
```

**Comment by lfrancke** [comment]
Thank you for this investigation and the detailed report.

I did review the initial code and I did stumble across the `ObjectMapper` instantiation. You're right: It's thread-safe and should be reused. In this case I didn't complain because I didn't think this class would be instantiated often. Sorry for the lazy question but you say it is "instantiated multiple times". Do you have numbers for that? I honestly would have thought it's instantiated once per process only.
Don't get me wrong: I believe you, I'd just like to understand what's going on and I won't be at a proper computer to check until next week.

And thanks on the HttpClient hint, that one didn't trigger for me. We'll be sure to take a look and will see if we can get a quick fix in for this in time for the next release. We'll discuss this next week.

**Comment by iAlex97** [comment]
I adjusted the logging a little bit to include the thread from where it was instantiated:

```java
LOG.debug("Started HdfsOpaAccessControlEnforcer on thread " + Thread.currentThread().getName());
```

On the active NameNode, it was instantiated from "IPC Service Handler" threads and definitely many times per same threads. I no longer have some logs, but I can reproduce them easily if needed.

**Comment by lfrancke** [comment]
Thank you! I'll check it out in the next couple of days. We're very close to our feature freeze for the next release but I'd really like to get some improvements merged for this as it's hurting....

**Comment by iAlex97** [comment]
Here are some logs:
```
2024-11-02 11:30:33,085 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 47 on default port 8020
2024-11-02 11:30:33,086 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 47 on default port 8020
2024-11-02 11:30:33,086 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 47 on default port 8020
2024-11-02 11:30:33,086 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 47 on default port 8020
2024-11-02 11:30:33,086 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 47 on default port 8020
...
2024-11-02 11:30:33,114 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 15 on default port 8020
2024-11-02 11:30:33,114 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 15 on default port 8020
2024-11-02 11:30:33,114 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 15 on default port 8020
2024-11-02 11:30:33,114 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 15 on default port 8020
...
2024-11-02 11:30:33,820 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 1 on default port 8020
2024-11-02 11:30:33,822 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 1 on default port 8020
...
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,887 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
2024-11-02 11:30:33,888 WARN  hadoop.StackableAccessControlEnforcer (StackableAccessControlEnforcer.java:<init>(79)) - Started HdfsOpaAccessControlEnforcer on thread IPC Server handler 27 on default port 8020
```

**Comment by iAlex97** [comment]
@lfrancke let me know if I can help you out with further testing or ideas.

**Comment by lfrancke** [comment]
So, I did look at it some more and found out that this is what the builder of HttpClient does when you do not specify your own executor:

```java
    Executor ex = builder.executor;
    if (ex == null) {
      ex = Executors.newCachedThreadPool(new DefaultThreadFactory(this.id));
      this.isDefaultExecutor = true;
    } else {
      this.isDefaultExecutor = false;
    }
```

In other words: It is the same code as you provide.
So, the real change you made was to move it to static so there is only one instance between all clients.

I'll put up a PR that should help by only ever creating a single instance of the Enforcer class in the first place.

**Comment by lfrancke** [comment]
I updated the description of this issue and the first PR is already merged: https://github.com/stackabletech/hdfs-utils/pull/50

**Comment by sbernauer** [comment]
I measured the performance gains of #50, #51 and #49 and combining all PRs reached the following numbers - which seems pretty good to me.

![Image](https://github.com/user-attachments/assets/987a1780-9fcb-4b05-9227-85d4697ce988)

I used `for i in $(seq 0 100); do echo "Creating $i" && bin/hdfs dfs -put -f /etc/hosts /bench/$i; done` to propagate some files and
`time bin/hdfs dfs -du -h /bench` as the benchmark (it does at least 100 calls to OPA)

**Comment by jradmacher** [comment]
> I used `for i in $(seq 0 100); do echo "Creating $i" && bin/hdfs dfs -put -f /etc/hosts /bench/$i; done` to propagate some files and `time bin/hdfs dfs -du -h /bench` as the benchmark (it does at least 100 calls to OPA)

What about a "deeper" path? If I remember my first debugging correctly, the message size exploded for a deep path.


>(it does at least 100 calls to OPA)

The amount of calls is an interesting question. `du` does at least two calls: One for the file path and one for the inode. The number of calls in total could be `1+2*100` or up to `(1+1+2*100) = (root+path+file+inode) *100` , if the path is not cached in any way.


Can you check, if a longer path changes your numbers significantly? 

`bin/hdfs dfs -mkdir -p /bench/deep/path/with/many/sub/sub/dirs/`
`for i in $(seq 0 100); do echo "Creating $i" && bin/hdfs dfs -put -f /etc/hosts /bench/deep/path/with/many/sub/sub/dirs/$i; done
time bin/hdfs dfs -du -h /bench`

**Comment by sbernauer** [comment]
Thanks for the feedback, numbers look good! (WARN log level - so no debugging)
![Image](https://github.com/user-attachments/assets/d8bbe071-91dd-47d2-87f9-52c5a8fa5788)

**Comment by lfrancke** [comment]
These improvements will ship in SDP 24.11 and I'll go ahead and close this issue.

## Release Notes

The performance of the https://docs.stackable.tech/home/stable/hdfs/usage-guide/security.html#_authorization[OPA Authorizer] has been greatly improved. This _can_ be a breaking change so please make sure to read the https://github.com/stackabletech/hdfs-utils/releases/tag/v0.4.0[release notes] of the hdfs-utils for details
